pkg/daemon/certificate_writer.go

```diff
@@ -102,37 +118,49 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
 	// Also annotate the latest config we've seen, so as to not write unnecessarily
 	currentNodeControllerConfigResource := dn.node.Annotations[constants.ControllerConfigResourceVersionKey]
 
-	if currentNodeControllerConfigResource != controllerConfig.ObjectMeta.ResourceVersion {
+	if currentNodeControllerConfigResource != controllerConfig.ObjectMeta.ResourceVersion || controllerConfig.Annotations["machineconfiguration.openshift.io/sync-service-ca"] == "true" {
 		pathToData := make(map[string][]byte)
 
 		kubeAPIServerServingCABytes := controllerConfig.Spec.KubeAPIServerServingCAData
 		cloudCA := controllerConfig.Spec.CloudProviderCAData
 		pathToData[caBundleFilePath] = kubeAPIServerServingCABytes
 		pathToData[cloudCABundleFilePath] = cloudCA
+		var data []byte
+		var err error
 
-		for bundle, data := range pathToData {
-			if !strings.HasSuffix(string(data), "\n") {
-				bString := string(data) + "\n"
-				data = []byte(bString)
-			}
-			if Finfo, err := os.Stat(bundle); err == nil {
-				var mode os.FileMode
-				Dinfo, err := os.Stat(filepath.Dir(bundle))
+		cm, cmErr := dn.kubeClient.CoreV1().ConfigMaps("openshift-machine-config-operator").Get(context.TODO(), "kubeconfig-data", v1.GetOptions{})
+		if cmErr != nil {
+			klog.Errorf("kubeconfig-data does not exist yet. %v", cmErr)
+		} else {
+			data, err = cmToData(cm, "ca-bundle.crt")
+			if err != nil {
+				klog.Errorf("kubeconfig-data ConfigMap not populated yet. %v", err)
+			} else if data != nil {
+				currentKC := clientcmdv1.Config{}
+				kcBytes, err := os.ReadFile(kubeConfigPath)
 				if err != nil {
-					mode = defaultDirectoryPermissions
-				} else {
-					mode = Dinfo.Mode()
-				}
-				// we need to make sure we honor the mode of that file
-				if err := writeFileAtomically(bundle, data, mode, Finfo.Mode(), -1, -1); err != nil {
 					return err
 				}
-			} else {
-				if err := writeFileAtomicallyWithDefaults(bundle, data); err != nil {
-					return err
+				if kcBytes != nil {
+					err = yaml.Unmarshal(kcBytes, &currentKC)
+					if err != nil {
+						return fmt.Errorf("Could not unmarshal kubeconfig into struct. Data: %s, Error: %v", string(kcBytes), err)
+					}
+					var newData []byte
+					if len(currentKC.Clusters) > 0 {
+						currentKC.Clusters[0].Cluster.CertificateAuthorityData = data
+						newData, err = json.Marshal(currentKC)
+						if err != nil {
+							return fmt.Errorf("Could not marshal kubeconfig into bytes. Error: %v", err)
+						}
+					}
+					pathToData[kubeConfigPath] = newData
 				}
 			}
 		}
+		if err := writeToDisk(pathToData); err != nil {
```

Comment:
suggestion: Make the name of this function more descriptive with respect to what data it's writing.

For example, `writeConfigMapToNodeFilesystem()` is a lot more descriptive than `writeToDisk()`.